x/batching/keeper/abci.go

```diff
@@ -0,0 +1,171 @@
+	curBatchNum, err := k.GetCurrentBatchNum(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+	dataLeaves, dataRootHex, err := k.ConstructDataResultTree(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+	valLeaves, valRootHex, err := k.ConstructValidatorTree(ctx)
+	if err != nil {
+		return types.Batch{}, err
+	}
+
+	return types.Batch{
+		BatchNumber:      curBatchNum,
+		BlockHeight:      ctx.BlockHeight(),
+		DataResultRoot:   dataRootHex,
+		ValidatorRoot:    valRootHex,
+		DataResultLeaves: dataLeaves,
+		ValidatorLeaves:  valLeaves,
+		BlockTime:        ctx.BlockTime(),
+	}, nil
+}
+
+// ConstructDataResultTree constructs a data result tree based on the
+// batching-ready data results returned from the core contract and
+// returns its leaves and its hex-encoded tree root.
+func (k Keeper) ConstructDataResultTree(ctx sdk.Context) ([][]byte, string, error) {
+	coreContract, err := k.wasmStorageKeeper.GetCoreContractAddr(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	// TODO: Deal with offset and limits. (#313)
+	queryRes, err := k.wasmViewKeeper.QuerySmart(ctx, coreContract, []byte(`{"get_data_results_by_status":{"status": "tallied", "offset": 0, "limit": 100}}`))
+	if err != nil {
+		return nil, "", err
+	}
+	if string(queryRes) == "[]" {
+		return nil, "", err
+	}
+
+	var dataResults []tallytypes.DataResult
+	err = json.Unmarshal(queryRes, &dataResults)
+	if err != nil {
+		return nil, "", err
+	}
+
+	leaves := make([][]byte, len(dataResults))
+	for _, res := range dataResults {
+		resHash, err := hex.DecodeString(res.ID)
+		if err != nil {
+			return nil, "", err
+		}
+		leaves = append(leaves, resHash)
+	}
+
+	curRoot := utils.RootFromEntries(leaves)
+	prevRoot, err := k.GetPreviousDataResultRoot(ctx)
+	if err != nil {
+		return nil, "", err
+	}
+	root := utils.RootFromLeaves([][]byte{prevRoot, curRoot})
+
+	// TODO update data result status on contract
+
+	return leaves, hex.EncodeToString(root), nil
+}
+
+type validatorData struct {
+	secp256k1PubKey cryptotypes.PubKey
+	power           math.Int // Validator power multiplied by 10^8.
+}
+
+// ConstructValidatorTree constructs a validator tree based on the
+// validators in the active set and their registered public keys.
+// It returns the tree's entries (unhashed leaf contents) and hex-encoded
+// root.
+func (k Keeper) ConstructValidatorTree(ctx sdk.Context) ([][]byte, string, error) {
+	var validators []validatorData
+	totalPower := math.ZeroInt()
+	err := k.stakingKeeper.IterateLastValidatorPowers(ctx, func(valAddr sdk.ValAddress, power int64) (stop bool) {
+		pubKey, err := k.pubKeyKeeper.GetValidatorKeyAtIndex(ctx, valAddr, utils.SEDAKeysIndexSecp256k1)
+		if err != nil {
+			if errors.Is(err, collections.ErrNotFound) {
+				return false
+			}
+			panic(err)
+		}
+		validators = append(validators, validatorData{
+			secp256k1PubKey: pubKey,
+			power:           math.NewInt(power).MulRaw(1e8),
+		})
+		totalPower = totalPower.AddRaw(power)
```

Comment:
There is total power on the staking keeper right?

Reply:
updated the code